torch/distributed/nn/api/remote_module.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/python3
+import uuid
+from typing import Any, Callable, Dict, Tuple
+
+import torch
+import torch.distributed.rpc as rpc
+from torch.distributed.nn.jit import instantiator
+
+
+def _gen_global_unique_name():
+    return f"{uuid.uuid4().hex}"
+
+
+def _instantiate_template(module_interface_cls, is_scriptable):
+    # Instantiate _RemoteModule class template on the local side.
+    generated_module = instantiator.instantiate_remote_module_template(
+        module_interface_cls, is_scriptable
+    )
+
+    remote_module_cls = generated_module._RemoteModule
+    remote_module_interface_cls = generated_module._RemoteModuleInterface
+
+    return remote_module_cls, remote_module_interface_cls
+
+
+def _script_module_creator_wrapper(module_creator, module_interface_cls, args, kwargs):
+    module = module_creator(*args, **kwargs)
+    script_module = torch.jit.script(module)
+    return rpc.RRef(script_module, module_interface_cls)
+
+
+def RemoteModule(
+    to: str,
+    module_creator: Callable,
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+    module_interface_cls: Any = None,
```

Comment:
I was wondering if we can do the following to avoid using `module_interface_cls`. We support only the following cases:

```
// 1. Use nn.Module directly.
RemoteModule(module_creator=nn.Linear())
NOTE: I think we can also easily support the case where module_creator itself is a type of ScriptModule.

// 2. Use nn.Module as part of function.
def my_module_creator(): -> nn.Module
  return MyModule()

RemoteModule(module_creator=my_module_creator())

// 3. Use nn.Module as ScriptModule.
def my_script_module_creator(): -> ScriptModule
  return torch.jit.script(MyModule())

RemoteModule(module_creator=my_script_module_creator())
```

In each of these cases, we can inspect either the class or the return type of the creator to determine what `module_interface_cls` should be. If the user doesn't specify a return type annotation, we just raise an error and ask them to provide one?

Also, I was chatting with @wanchaol offline to see if we can come up with options to make this cleaner. I think it might make sense to have a 30min meeting next week to discuss a few ideas around this.

Reply:
RemoteModule is relying on local RRef constructor API. 

A separate PR later should clean up local RRef constructor API and wrap type info when passing and pickling RRefs.

This will need careful design.